csrc/id_model/id_graph.cpp

```diff
@@ -0,0 +1,1016 @@
+  IdGroups producers;
+  for (auto expr : std::vector<Expr*>{expr0, expr1}) {
+    for (auto input_id : ir_utils::filterByType<IterDomain>(expr->inputs())) {
+      producers.pushBack(toGroup(input_id));
+    }
+  }
+
+  for (auto producer_group : producers) {
+    uniqueUses().at(producer_group).erase(expr0_orig_group);
+    uniqueUses().at(producer_group).erase(expr1_orig_group);
+    uniqueUses().at(producer_group).pushBack(expr_new_group);
+  }
+
+  // Update unique definitinos of consumers
+  IdGroups consumers;
+  for (auto expr : std::vector<Expr*>{expr0, expr1}) {
+    for (auto output_id : ir_utils::filterByType<IterDomain>(expr->outputs())) {
+      consumers.pushBack(toGroup(output_id));
+    }
+  }
+
+  for (auto consumer_group : consumers) {
+    uniqueDefinitions().at(consumer_group).erase(expr0_orig_group);
+    uniqueDefinitions().at(consumer_group).erase(expr1_orig_group);
+    uniqueDefinitions().at(consumer_group).pushBack(expr_new_group);
+  }
+}
+
+bool IdGraph::mapThroughExpr(Expr* first, Expr* second, bool forward) {
+  if (first == nullptr || second == nullptr) {
+    return false;
+  }
+
+  if (!exprsMap(first, second, forward)) {
+    return false;
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      propagate_exprs_,
+      "Asked to propagate expression mappings on a graph that has propagate_exprs_ disabled.");
+
+  auto first_ids = ir_utils::filterByType<IterDomain>(
+                       forward ? first->outputs() : first->inputs())
+                       .vector();
+  auto second_ids = ir_utils::filterByType<IterDomain>(
+                        forward ? second->outputs() : second->inputs())
+                        .vector();
+  TORCH_INTERNAL_ASSERT(
+      first_ids.size() == second_ids.size(),
+      "This should be unreachable, if transformation expressions match, their number of inputs and outputs should as well.\n However found:\n",
+      first->toString(),
+      "\nand\n",
+      second->toString());
+  for (auto out_i : c10::irange(first_ids.size())) {
+    mapIds(first_ids[out_i], second_ids[out_i]);
+  }
+
+  return true;
+}
+
+void IdGraph::mapThroughLoopSwizzles() {
+  std::vector<Swizzle2D*> all_swizzles;
+
+  for (auto expr_set : disjointExprSets().disjointSets()) {
+    auto swizzles_in_expr_set = ir_utils::filterByType<Swizzle2D>(
+        expr_set->vector().begin(), expr_set->vector().end());
+    all_swizzles.insert(
+        all_swizzles.end(),
+        swizzles_in_expr_set.begin(),
+        swizzles_in_expr_set.end());
+  }
+
+  for (auto swizzle : all_swizzles) {
+    if (swizzle->swizzleMode() == SwizzleMode::Loop) {
+      mapIds(swizzle->inX(), swizzle->outX());
+      mapIds(swizzle->inY(), swizzle->outY());
+    }
+  }
+}
+
+void IdGraph::mapThroughTrivialExprs() {
+  // Grab all expressions
+  std::vector<Expr*> exprs;
+
+  for (auto expr_group : disjointExprSets().disjointSets()) {
+    for (auto expr : *expr_group) {
+      exprs.push_back(expr);
+    }
+  }
```

Comment:
I am confused about when we should iterate over all exprs and when we should iterate over only the group. For example, for this, if we just do `exprs.push_back(expr_group.front())`, would it work?